llvm-build/utils.ps1

```diff
@@ -0,0 +1,278 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+if (Test-Path function:\exec) {
+    # already included
+    return
+}
+
+if (!(Test-Path function:\Get-RepoRoot)) {
+    # git revparse uses cwd. E2E builds use a different
+    # working dir, so we pin it to out repo (submodule in E2E)
+    function Get-RepoRoot {
+        exec -wd $PSScriptRoot {
+            git rev-parse --show-toplevel
+        }
+    }
+}
+
+# Fix temp path for non-windows platforms if missing
+if (!(Test-Path env:\TEMP)) {
+    $env:TEMP = [System.IO.Path]::GetTempPath()
+}
+
+# Test whether we build the compiler components.
+function Test-BuildCompiler {
+    # By default we want to build the compiler
+    # If no env var is defined we default to true
+    if (!(Test-Path env:\ENABLE_COMPILER)) {
+        $true
+    }
+    else {
+        # The env var exists, we use its value
+        # Coerce falsy values and negate to keep the defalt $true
+        $env:ENABLE_COMPILER -ne $false
+    }
+}
+
+# Test whether we build the VSIX components.
+function Test-BuildVsix {
+    ($Env:ENABLE_VSIX -ne "false") -and $IsWindows
+}
+
+# Test whether we build the LLVM components.
+function Test-BuildLlvmComponents {
+    ((Test-Path env:\ENABLE_LLVM_BUILDS) -and ($env:ENABLE_LLVM_BUILDS -eq $true))
+}
+
+####
+# Utilities
+####
+
+# executes the sciptblock. If the working directory is specitied, the block is
+# exeucted in that directory. If the $LASTEXITCODE from the block is not 0,
+# an exception is thrown.
+function exec {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        [scriptblock]$cmd,
+        [string]$errorMessage = "Failed to execute command",
+        [Alias("wd")]
+        [string]$workingDirectory = $null
+    )
+    try {
+
+        if ($workingDirectory) {
+            Push-Location -Path $workingDirectory
+        }
+
+        $global:lastexitcode = 0
+        & $cmd
+        if ($global:lastexitcode -ne 0) {
+            throw "exec: $errorMessage"
+        }
+    }
+    finally {
+        if ($workingDirectory) {
+            Pop-Location
+        }
+    }
+}
+
+# tests whether a condition is true. Throws an exception with the specified error message
+# if the condition check fails.
+function Assert {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        $conditionToCheck,
+
+        [Parameter(Mandatory = $true)]
+        [string]$failureMessage
+    )
+
+    if (-not $conditionToCheck) {
+        throw ('Assert: {0}' -f $failureMessage)
+    }
+}
+
+# returns true if the script is running on a build agent, false otherwise
+function Test-CI {
+    if (Test-Path env:\TF_BUILD) {
+        $true
+    }
+    elseif ((Test-Path env:\CI)) {
+        $env:CI -eq $true
+    }
+    else {
+        $false
+    }
+}
```

Comment:
Nit: I personally prefer the functions returning Boolean to be named such that from the name it is immediately clear
* that the function returns Boolean,
* what it means if `true` is returned and what if `false` is returned.

Currently the names `Test-BuildVsix`, `Test-BuildLlvmComponents`, `Test-CI` do not tell me that. These names tell me that some sets of tests are executed for the Vsix Build, for Llvm Components Build, and for CI.

You may want to consider changing to something like '**Should**BuildVsix', '**Must**BuildLlvmComponents, '**Running**OnCI'.

Reply:
`Test-` is the idiomatic prefix for PowerShell commands that return streams of boolean outputs: 

<img width="660" alt="image" src="https://user-images.githubusercontent.com/31516/148878257-a86964fb-d85f-4dc8-ac4c-9341e6e3dc08.png">

https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.2

Reply:
Thanks!